crates/engine/tree/src/chain.rs

```diff
@@ -0,0 +1,142 @@
+///
+///
+/// ## Control flow
+///
+/// The [`ChainOrchestrator`] is responsible for controlling the pipeline sync and additional hooks.
+/// It polls the given `handler`, which is responsible for advancing the chain, how is up to the
+/// handler. However, due to database restrictions (e.g. exclusive write access), following
+/// invariants apply:
+///  - If the handler requests a pipeline run (e.g. [`PipelineAction::SyncPipeline`]), the handler
+///    must ensure that while the pipeline is running, no other write access is granted.
+///  - At any time the [`ChainOrchestrator`] can request exclusive write access to the database
+///    (e.g. if pruning is required), but will not do so until the handler has acknowledged the
+///    request for write access.
+///
+/// The [`ChainOrchestrator`] polls the [`ChainHandler`] to advance the chain and handles the
+/// emitted events. Requests and events are passed to the [`ChainHandler`] via
+/// [`ChainHandler::on_event`].
+#[must_use = "Stream does nothing unless polled"]
+pub struct ChainOrchestrator<T>
+where
+    T: ChainHandler,
+{
+    /// The handler for advancing the chain.
+    handler: T,
+    /// Controls pipeline sync.
+    pipeline: (),
+    /// Additional hooks (e.g. pruning) that can require exclusive access to the database.
+    hooks: (),
+}
+
+impl<T> ChainOrchestrator<T>
+where
+    T: ChainHandler,
+{
+    /// Returns the handler
+    pub const fn handler(&self) -> &T {
+        &self.handler
+    }
+
+    /// Returns a mutable reference to the handler
+    pub fn handler_mut(&mut self) -> &mut T {
+        &mut self.handler
+    }
+
+    /// Internal function used to advance the chain.
+    ///
+    /// Polls the `ChainOrchestrator` for the next event.
+    #[tracing::instrument(level = "debug", name = "ChainOrchestrator::poll", skip(self, cx))]
+    fn poll_next_event(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<ChainEvent> {
+        todo!("do we need this?")
+    }
+}
+
+impl<T> Stream for ChainOrchestrator<T>
+where
+    T: ChainHandler,
+{
+    type Item = ChainEvent;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.as_mut().poll_next_event(cx).map(Some)
+    }
+}
+
+/// Event emitted by the [`ChainOrchestrator`]
+///
+/// These are meant to be used for observability and debugging purposes.
+pub enum ChainEvent {
+    /// Pipeline sync started
+    PipelineStarted,
+}
+
+/// A trait that advances the chain by handling actions.
+///
+/// This is intended to be implement the chain consensus logic, for example `engine` API.
+pub trait ChainHandler: Send + Sync {
+    /// Informs the handler about an event from the [`ChainOrchestrator`].
+    fn on_event(&mut self, event: FromOrchestrator);
+
+    /// Polls for actions that [`ChainOrchestrator`] should handle.
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<HandlerEvent>;
+}
```

Comment:
This is really good because it completely generalizes between a handler that's for Engine API For L1 usage, an optimized Engine API Handler for OP Sequencer usage which does FCU->GP->NP in a loop instead of allowing forking, or a handler for usage with ABCI or other consensus interfaces.